### PR TITLE
Tests separation

### LINKED ISSUE #949: "email del" does not remove data directory, even if it says so
## Context
While investigating my failing test on my PR (https://github.com/tomav/docker-mailserver/pull/884), I discovered that one feature is not working as expected and its test fails to provide guard.

When running `./setup -c mail email del a.user@server.com`, there is an option to delete the data folder.

Sadly when that option is selected (-y or y at run time), it does not delete the directory.

## Expected Behavior

The user should be removed from the account file and if I wish the maildir folder should be deleted too.

## Actual Behavior

Mail folder is not deleted.

## Possible Fix

The script delmailuser is ran from setup.sh in a docker "run" command. This command only mount the configuration directory and not the persistence of the mail folder so obviously the "normal" server will not have its maildir (inside the container or on the host filesystem) deleted as it only tries to delete a folder in a docker run once container.

## Steps to Reproduce

1. launch the setup for tests "mail, mail-xx"
2. add a user through the script
3. delete a user through the script
4. check the existence of the folder on the main "mail" container

## The impact

The test that is supposedly validating this feature does not its job and fails to "see" the issues.

I will comment that test in my PR as it does not work as expected (when I finish)

### PR DESCRIPTION
**Pull request by 17Halbe**
**This is more a test for Travis than the final product.** 

- introduced multithreaded container startup
- separated test cases according to the container they are run on
- Introduced optional `PERF` option for `make`. (The higher the PERF number, the lower the sleep time in the Makefile)
- Introduced optional `CONTAINER` option for `make` to decrease development and testing time. Specific Container can be chosen to be tested. It is provided as a space separated list of container names. List of all Container:`CONTAINER=from_image mail mail_privacy mail_pop3 mail_smtponly mail_smtponly_without_config mail_override_hostname mail_fail2ban mail_fetchmail mail_disabled_clamav_spamassassin mail_manual_ssl mail_with_ldap mail_postscreen mail_lmtp_ip mail_with_postgrey mail_undef_spam_subject mail_with_relays mail_with_imap`
- added a `clean` prerequisite prior building to the main/default targets. One would want to get everything reset before commencing a new test.
- cleaned up obsolete setup.sh arguments (-c CONTAINER for functions running via the image)
- made some tests less flexible by replacing the dynamic image name lookup with a fixed value, to make them independent from the running mail container)
- Fixes userdel bug (closes #949)

The number of tests didn't change:
screenshot from 2018-04-19 09-34-53

Right now the tests aren't multithreaded. There are quite a lot of problems regarding processor load, timing and displaying the results. 
`tests.bats` is still included for reference. It might look a bit unordered, but it's the final version I separated the tests from.
See #945
Concerns #950

### CODE REVIEW

**Comment by 17Halbe** [comment]
Ok this seems to work.
Todo: 
- Parallizing tests
- Separating fixtures into container specific targets
- Creating Test/Container specific config directories
- What else?

File: Makefile [hunk 1]
```diff
@@ -218,47 +240,76 @@ run:
...
-	docker exec mail /bin/sh -c "nc 0.0.0.0 25 < /tmp/docker-mailserver-test/email-templates/existing-user-and-cc-local-alias.txt"
-	docker exec mail /bin/sh -c "nc 0.0.0.0 25 < /tmp/docker-mailserver-test/email-templates/existing-regexp-alias-external.txt"
-	docker exec mail /bin/sh -c "nc 0.0.0.0 25 < /tmp/docker-mailserver-test/email-templates/existing-regexp-alias-local.txt"
-	docker exec mail /bin/sh -c "nc 0.0.0.0 25 < /tmp/docker-mailserver-test/email-templates/existing-catchall-local.txt"
-	docker exec mail /bin/sh -c "nc 0.0.0.0 25 < /tmp/docker-mailserver-test/email-templates/sieve-spam-folder.txt"
-	docker exec mail /bin/sh -c "nc 0.0.0.0 25 < /tmp/docker-mailserver-test/email-templates/sieve-pipe.txt"
-	docker exec mail /bin/sh -c "nc 0.0.0.0 25 < /tmp/docker-mailserver-test/email-templates/non-existing-user.txt"
-	docker exec mail_disabled_clamav_spamassassin /bin/sh -c "nc 0.0.0.0 25 < /tmp/docker-mailserver-test/email-templates/existing-user1.txt"
+	$(eval NEEDSSLEEP=false)
+  ifneq (,$(filter mail,$(CONTAINER)))
```
Comment (crazystick) [inline Makefile]: Just wondering, why not use the same fixtures_<container> pattern like you use for build_<container>?

File: Makefile [hunk 2]
```diff
@@ -218,47 +240,76 @@ run:
...
+  ifneq (,$(filter mail_privacy,$(CONTAINER)))
+		docker exec mail_privacy /bin/sh -c "openssl s_client -quiet -starttls smtp -connect 0.0.0.0:587 < /tmp/docker-mailserver-test/email-templates/send-privacy-email.txt"
+		$(eval NEEDSSLEEP=true)
+  endif
+  ifneq (,$(filter mail_override_hostname,$(CONTAINER)))
+		docker exec mail_override_hostname /bin/sh -c "nc 0.0.0.0 25 < /tmp/docker-mailserver-test/email-templates/existing-user1.txt"
+		$(eval NEEDSSLEEP=true)
+  endif
+  # Wait for mails to be analyzed
+	@if [ $(NEEDSSLEEP) = true ]; then \
```
Comment (crazystick) [inline Makefile]: Sleeps should just be part of the fixture with duration defined by the test author - if I make one fixture to deliver one email, I don't need to sleep for 120s if I'm just running my own fixture.

File: Makefile (hunk 1 above)
Comment (crazystick) [reply Makefile]: Ah - is that your TODO item 2?

File: Makefile (hunk 2 above)
Comment (17Halbe) [reply Makefile]: You are right and that‘s on the todo list.

**Comment by 17Halbe** [comment]
Ok, did some more testing, but really don't find the time right now. Unfortunately the tests are as or even less reliable than before. 
Those sleep commands are a real problem. Their usefulness really depends on the available resources of the underlying hardware. 
So how do we figure out, if a container is up and running?
- finding its pid and checking for its resource usage,
- greping the log for a "up'n'running" message from the container(after startup and after test-message processing.(Probably via [postqueue](http://www.postfix.org/postqueue.1.html)?))

Does anyone has a bit more experience and ideas regarding that issue?

**Comment by johansmitsnl** [comment]
@17Halbe the issue is indeed how to detect if a container is running.
Even when you add to the log "running" dovecot can still be generating ssl keys or whatever the services do when they are started from the main script until they are ready.
Maybe you can open a tcp connection and check for certain output in a while loop of Postfix and Dovecot.

If this works then send the test data into the container, wait for processing (post queue sounds good) and then start the actual tests.

**Comment by 17Halbe** [comment]
Sorry for the "long time no see". I was sick and away for it. I won't be able to contribute for some time. Sorry. Feel free to close this one.

**Comment by erik-wramner** [comment]
While this is outdated the changes as such would be extremely valuable. I think we should split it into smaller parts and get them into the master branch one change at a time. I'll leave this open for now.

**Comment by erik-wramner** [comment]
Closing as similar work is under way in other PRs.

**Pull request closed without merging**